pkg/services/accesscontrol/permreg/permreg.go

```diff
@@ -0,0 +1,184 @@
+		}
+		// Add the most specific format kind:attribute:*
+		acceptedPrefixesList = append(acceptedPrefixesList, prefix+"*")
+	}
+	return acceptedPrefixesList
+}
+
+type PermissionRegistry interface {
+	RegisterPluginScope(scope string)
+	RegisterPermission(action, scope string)
+	IsPermissionValid(action, scope string) error
+	GetScopePrefixes(action string) (PrefixSet, bool)
+}
+
+type PrefixSet map[string]bool
+
+var _ PermissionRegistry = &permissionRegistry{}
+
+type permissionRegistry struct {
+	actionScopePrefixes map[string]PrefixSet // TODO use thread safe map
+	kindScopePrefix     map[string]string
+	logger              log.Logger
+}
+
+func ProvidePermissionRegistry() PermissionRegistry {
+	return newPermissionRegistry()
+}
+
+func newPermissionRegistry() *permissionRegistry {
+	// defaultKindScopes maps the most specific accepted scope prefix for a given kind (folders, dashboards, etc)
+	defaultKindScopes := map[string]string{
+		"teams":           "teams:id:",
+		"users":           "users:id:",
+		"datasources":     "datasources:uid:",
+		"dashboards":      "dashboards:uid:",
+		"folders":         "folders:uid:",
+		"annotations":     "annotations:type:",
+		"apikeys":         "apikeys:id:",
+		"orgs":            "orgs:id:",
+		"plugins":         "plugins:id:",
+		"provisioners":    "provisioners:",
+		"reports":         "reports:id:",
+		"permissions":     "permissions:type:",
+		"serviceaccounts": "serviceaccounts:id:",
+		"settings":        "settings:",
+		"global.users":    "global.users:id:",
+		"roles":           "roles:uid:",
+		"services":        "services:",
+	}
+	return &permissionRegistry{
+		actionScopePrefixes: make(map[string]PrefixSet, 200),
+		kindScopePrefix:     defaultKindScopes,
+		logger:              log.New("accesscontrol.permreg"),
+	}
+}
+
+func (pr *permissionRegistry) RegisterPluginScope(scope string) {
+	if scope == "" {
+		return
+	}
+
+	scopeParts := strings.Split(scope, ":")
+	// If the scope is already registered, return
+	if _, found := pr.kindScopePrefix[scopeParts[0]]; found {
+		return
+	}
+
+	// If the scope contains an attribute part, register the kind and attribute
+	if len(scopeParts) > 2 {
+		kind, attr := scopeParts[0], scopeParts[1]
+		pr.kindScopePrefix[kind] = kind + ":" + attr + ":"
+		pr.logger.Debug("registered scope prefix", "kind", kind, "scope_prefix", kind+":"+attr+":")
+		return
+	}
+
+	pr.logger.Debug("registered scope prefix", "kind", scopeParts[0], "scope_prefix", scopeParts[0]+":")
+	pr.kindScopePrefix[scopeParts[0]] = scopeParts[0] + ":"
+}
+
+func (pr *permissionRegistry) RegisterPermission(action, scope string) {
+	if _, ok := pr.actionScopePrefixes[action]; !ok {
+		pr.actionScopePrefixes[action] = PrefixSet{}
+	}
+
+	if scope == "" {
+		// scopeless action
+		return
+	}
+
+	kind := strings.Split(scope, ":")[0]
+	scopePrefix, ok := pr.kindScopePrefix[kind]
+	if !ok {
+		// TODO (gamab) should we register the scope prefix anyway?
```

Comment:
what are the implications of doing vs not doing?

doing:
- may introduce edge cases if attempting to retrieve all scope prefixes in a feature
  - is it done?

not doing:
- does it introduce the risk of sec vulns since the overlap would also be an unknown scope prefix?

Reply:
Currently, the permission registry relies on the hard coded list of `kindScopePrefixes` that contains the most specific scope possible in order to detect invalid permissions upon role creation or update (ex: `users:uid:admin` when we expect `users:id:` scopes).

**Why didn't we make this dynamic in the first place?**

I assume this is because our fixed roles' permissions are usually short wildcards (ex: `users:*` instead of `users:id:*`) and we miss the attribute part of the scope to populate `kindScopePrefixes`.

**Implication of not doing (register the scope prefix anyway)**

If we introduce a **new resource scope** (ex: `bananas:uid:*`) we will log this **warning** and the permission registry won't have this scope. It will also wrongly **consider the associated actions to be scopeless** (ex: `bananas:read`; `bananas:write`).

This currently won't have an impact on plugins (as we only allow `datasources` and `plugins` based actions which are in the hard coded list). It will just not work if plugins can protect their includes/routes with a new kind of actions (ex: `banana:read`).

No security risk

**Implication of doing (register the scope prefix anyway)**

In `kindScopePrefixes` we will potentially have a scope prefix that lacks the attribute part (ex: `bananas:` instead of `bananas:uid:`) and therefore **might consider an invalid scope as valid** (ex: `bananas:id:2`)
This currently won't have an impact on plugins (as we only allow `datasources` and `plugins` based actions which are in the hard coded list). Even if plugins can protect their includes/routes with a new kind of actions (ex: `banana:read`), we can still generate the right scope (ex: `bananas:uid`).

No security risk

**Alternative**

Instead of logging a warning, we can error out. This will make it less likely that we forget to populate `kindScopePrefixes`.

Reply:
> I assume this is because our fixed roles' permissions are usually short wildcards (ex: users:* instead of users:id:*) and we miss the attribute part of the scope to populate kindScopePrefixes.

Good point, that makes sense.

>Instead of logging a warning, we can error out. This will make it less likely that we forget to populate kindScopePrefixes.

I think that actually makes sense - this would not happen for plugins, as we call `RegisterPluginScope`, and it should not happen for any new fixed roles (it would be caught immediately upon Grafana startup, so would be caught during testing).